src/tests/rpp/test_flat_map.cpp

```diff
@@ -0,0 +1,126 @@
+            {
+                CHECK(mock.get_total_on_next_count() == 0);
+                CHECK(mock.get_on_completed_count() == 0);
+                CHECK(mock.get_on_error_count() == 1);
+            }
+        }
+        SECTION("subscribe using flat_map with empty")
+        {
+            obs | rpp::operators::flat_map([](int) { return rpp::source::empty<int>(); })
+                | rpp::ops::subscribe(mock.get_observer());
+            SECTION("observer obtains values from underlying observables")
+            {
+                CHECK(mock.get_total_on_next_count() == 0);
+                CHECK(mock.get_on_completed_count() == 1);
+                CHECK(mock.get_on_error_count() == 0);
+            }
+        }
+        SECTION("subscribe using flat_map with never")
+        {
+            obs | rpp::operators::flat_map([](int) { return rpp::source::never<int>(); })
+                | rpp::ops::subscribe(mock.get_observer());
+            SECTION("observer obtains values from underlying observables")
+            {
+                CHECK(mock.get_total_on_next_count() == 0);
+                CHECK(mock.get_on_completed_count() == 0);
+                CHECK(mock.get_on_error_count() == 0);
+            }
+        }
+        SECTION("subscribe using flat_map with empty in middle")
+        {
+            obs | rpp::operators::flat_map([](int v) {
+                    if (v == 2) 
+                        return rpp::source::empty<int>().as_dynamic();
+                    return rpp::source::just(v).as_dynamic(); 
+                  })
+                | rpp::ops::subscribe(mock.get_observer());
+            SECTION("observer obtains values from underlying observables")
+            {
+                CHECK(mock.get_received_values() == std::vector{ 1, 3 });
+                CHECK(mock.get_on_completed_count() == 1);
+                CHECK(mock.get_on_error_count() == 0);
+            }
+        }
+        SECTION("subscribe using flat_map with error in middle")
+        {
+            obs | rpp::operators::flat_map([](int v) {
+                    if (v == 2) 
+                        return rpp::source::error<int>(std::make_exception_ptr(std::runtime_error{""})).as_dynamic();
+                    return rpp::source::just(v).as_dynamic(); 
+                  })
+                | rpp::ops::subscribe(mock.get_observer());
+            SECTION("observer obtains values from underlying observables")
+            {
+                CHECK(mock.get_total_on_next_count() == 0);
+                CHECK(mock.get_on_completed_count() == 0);
+                CHECK(mock.get_on_error_count() == 1);
+            }
+        }
+    }
+    /*
+    SECTION("flat_map doesn't produce extra copies")
+    {
+        copy_count_tracker verifier{};
+        auto obs = rpp::source::just<TestType>(std::move(verifier))
+                 | rpp::ops::flat_map([](copy_count_tracker&& verifier) { return rpp::source::just<TestType>(std::move(verifier)); });
```

Comment:
@victimsnino do you know why sources don't move objects through operators? Preventing me from taking verifier as an rvalue ref here

Reply:
Because it would be moved IF possible, but no guarantees. If it would be only pure rvalue, then it means second repeated subscribe would see nothing :)

Reply:
I see, I added some cases for copies then

Reply:
if you would use create with std::move inside, then it would obtain value by rvalue reference